README.md
# NAME

Git::Lint - lint git commits and messages

# SYNOPSIS

    use Git::Lint;

    my $lint = Git::Lint->new();
    $lint->run({ check => 'commit', profile => 'default' });
    $lint->run({ check => 'message', file => 'file_path', profile => 'default' });

    git-lint [--check commit] [--check message <message_file>]
             [--profile <name>]
             [--version] [--help]

# DESCRIPTION

`Git::Lint` is a pluggable framework for linting git commits and messages.

For the commandline interface to `Git::Lint`, see the documentation for [git-lint](https://metacpan.org/pod/git-lint).

For adding check modules, see the documentation for [Git::Lint::Check::Commit](https://metacpan.org/pod/Git%3A%3ALint%3A%3ACheck%3A%3ACommit) and [Git::Lint::Check::Message](https://metacpan.org/pod/Git%3A%3ALint%3A%3ACheck%3A%3AMessage).

# CONSTRUCTOR

- new

    Returns a reference to a new `Git::Lint` object.

# METHODS

- run

    Loads the check modules as defined by `profile`.

    `run` expects the following arguments:

    **profile**

    The name of a defined set of check modules to run.

    **check**

    Either `commit` or `message`.

    **file**

    If `check` is `message`, `file` is required.

- config

    Returns the [Git::Lint::Config](https://metacpan.org/pod/Git%3A%3ALint%3A%3AConfig) object created by `Git::Lint`.

# INSTALLATION

To install `Git::Lint`, download the latest release, then extract.

    tar xzvf Git-Lint-0.008.tar.gz
    cd Git-Lint-0.008

or clone the repo.

    git clone https://github.com/renderorange/Git-Lint.git
    cd Git-Lint

Generate the build and installation tooling.

    perl Makefile.PL

Then build, test, and install.

    make
    make test && make install

`Git::Lint` can also be installed using [cpanm](https://metacpan.org/pod/cpanm).

    cpanm Git::Lint

# CONFIGURATION

Configuration is done through `git config` files (`~/.gitconfig` or `/repo/.git/config`).

Only one profile, `default`, is defined internally. `default` contains all check modules by default.

The `default` profile can be overridden through `git config` files (`~/.gitconfig` or `/repo/.git/config`).

To set the default profile to only run the `Whitespace` commit check:

    [lint "profiles.commit"]
        default = Whitespace

Or set the default profile to `Whitespace` and the fictional commit check, `Flipdoozler`:

    [lint "profiles.commit"]
        default = Whitespace, Flipdoozler

Additional profiles can be added with a new name and list of checks to run.

    [lint "profiles.commit"]
        default = Whitespace, Flipdoozler
        hardcore = Other, Module, Names

Message check profiles can also be defined.

    [lint "profiles.message"]
        # override the default profile to only contain SummaryLength, SummaryEndingPeriod, and BlankLineAfterSummary
        default = SummaryLength, SummaryEndingPeriod, BlankLineAfterSummary
        # create a summary profile with specific modules
        summary = SummaryEndingPeriod, SummaryLength

An example configuration is provided in the `examples` directory of this project.

Configuration is required.  If no configuration exists, an error will be printed to STDERR, but the action allowed to complete.

    blaine@base ~/git/test (master *) $ git add test
    blaine@base ~/git/test (master +) $ git commit
    git-lint: [error] configuration setup is required. see the documentation for instructions.
    [master 894b6d0] test
     1 file changed, 1 insertion(+), 1 deletion(-)
    blaine@base ~/git/test (master) $

# ADDING NEW CHECK MODULES

`git-lint` can be configured to load check modules from a local directory using the `localdir` configuration setting.

To load modules from a local directory, add the lint `config` setting, with `localdir` key and directory location to the git config file.

    [lint "config"]
        localdir = /home/blaine/tmp/git-lint/lib

In this example, we're adding a new commit check, `Flipdoozler`.  Create the local directory and path for the new module.

    $ mkdir -p /home/blaine/tmp/git-lint/lib/Git/Lint/Check/Commit

Then add the new check module.

    $ vi /home/blaine/tmp/git-lint/lib/Git/Lint/Check/Commit/Flipdoozler.pm
    package Git::Lint::Check::Commit::Flipdoozler;
    ...

Update the commit check profile to use the new module.

    [lint "profiles.commit"]
        default = Whitespace, IndentTabs, MixedIndentTabsSpaces, Flipdoozler

`git-lint` will now warn for the check contained in Flipdoozler.

    blaine@base ~/git/test (master +) $ git commit
    git-lint: [commit] test - Flipdoozler (line 18)
    blaine@base ~/git/test (master +) $

# ENABLING CHECKS FOR REPOS

To enable as a `pre-commit` hook, copy the `pre-commit` script from the `example/hooks` directory into the `.git/hooks` directory of the repo you want to check.

Once copied, update the path and options to match your path and preferred profile.

To enable as a `commit-msg` hook, copy the `commit-msg` script from the `example/hooks` directory into the `.git/hooks` directory of the repo you want to check.

# COPYRIGHT AND LICENSE

Copyright (c) 2022 Blaine Motsinger under the MIT license.

# AUTHOR

Blaine Motsinger `blaine@renderorange.com`


Changes
Revision history for perl module Git::Lint

0.018 2023-12-27
 - Add commit checks IndentSpaces and IndentTabs
 - Add known caveats section to the README

0.017 2023-12-12
 - Add commit check MixedIndentTabsSpaces

0.016 2022-08-08
 - Remove 000_load.t test

0.015 2022-08-07
 - Rename AUTHOR tests for earlier testing order
 - Add 200_manifest.t test
 - Add unpacked dist dir to MANIFEST.SKIP
 - Add unpacked dist dir to .gitignore
 - Add 201_version.t test

0.014 2022-03-05
 - Add config.t to MANIFEST

0.013 2022-03-05
 - Correct path for hook examples
 - Add test for Git::Lint config
 - Build profiles hashref regardless of found modules

0.012 2022-03-02
 - Correct missing changelog entry

0.011 2022-03-01
 - Update documentation
 - Add argument verification for run method

0.010 2022-02-28
 - Override user_config in new.t
 - Add CONTRIBUTING.md file
 - Move _config out of the Git::Lint object

0.009 2022-02-26
 - Fix documentation for message check modules
 - Fix double installation sections in documentation
 - Add cpanm to installation instructions
 - Rename Config new to load
 - Rename report method to format_issue
 - Update MANIFEST for report method rename
 - Remove GoFmt commit check module
 - Update MANIFEST for GoFmt removal
 - Fix message test for non-English locales
 - Remove Config new test from MANIFEST

0.008 2022-02-25
 - Move config and hook examples into examples dir
 - Documentation update
 - Add initial build and install tooling

0.007 2022-02-25
 - Simplify issue output

0.006 2022-02-25
 - Add opt message output if not filename
 - Remove incorrect exit in diff method
 - Add BlankLineAfterSummary.pm message check module
 - Add SummaryEndingPeriod.pm message check module
 - Add check message tests
 - Remove t/data from gitignore
 - Add tests for Check::Message->message
 - Add input verification for Check::Message->message

0.005 2022-02-20
 - Update usage output
 - Update documentation
 - Add body line length message check
 - Rename Summary check to SummaryLength
 - Add documentation for check modes
 - Add description var to GoFmt.pm

0.004 2022-02-16
 - Add GoFmt.pm check module
 - Add the get_filename method
 - Add MIT license

0.003 2022-02-13
 - Move loader out of new method
 - Add message check mode
 - Add commit-msg hook
 - Update docs and pod
 - Update pre-commit hook for new syntax
 - Fix Config/new.t test for message check changes
 - Update perlcritic config
 - Update gitignore

0.002 2020-09-13
 - Add defined loop var instead of internal var
 - Skip lines denoting the files being diff'd
 - Add program name to output
 - Add documentation to the README

0.001 2020-06-30
 - Initial version


lib/Git/Lint.pm
package Git::Lint;

use strict;
use warnings;

use Git::Lint::Config;
use Try::Tiny;
use Module::Loader;

our $VERSION = '0.018';

my $config;

sub new {
    my $class = shift;
    my $self  = { issues => undef };

    bless $self, $class;

    $config = Git::Lint::Config->load();

    return $self;
}

sub config {
    my $self = shift;
    return $config;
}

sub run {
    my $self = shift;
    my $opt  = shift;

    foreach my $required (qw{profile check}) {
        die "$required is required\n"
            unless defined $opt->{$required};
    }

    if ( $opt->{check} ne 'message' && $opt->{check} ne 'commit' ) {
        die "check must be either message or commit\n";
    }

    if ( $opt->{check} eq 'message' ) {
        die "file is required if check is message\n"
            unless defined $opt->{file};
    }

    die 'profile ' . $opt->{profile} . ' was not found' . "\n"
        unless exists $self->config->{profiles}{ $opt->{check} }{ $opt->{profile} };

    my $check = lc $opt->{check};
    $check = ucfirst $check;

    my $loader = Module::Loader->new;

    my @issues;
    foreach my $module ( @{ $self->config->{profiles}{ $opt->{check} }{ $opt->{profile} } } ) {
        my $class = q{Git::Lint::Check::} . $check . q{::} . $module;
        try {
            $loader->load($class);
        }
        catch {
            my $exception = $_;
            die "$exception\n";
        };
        my $plugin = $class->new();

        # TODO: this would be better to be named the same method
        my $input = ( $opt->{check} eq 'commit' ? $class->diff() : $class->message( file => $opt->{file} ) );

        # ensure the plugins don't manipulate the original input
        my @lines = @{$input};
        push @issues, $plugin->check( \@lines );
    }

    foreach my $issue (@issues) {
        if ( $opt->{check} eq 'commit' ) {
            push @{ $self->{issues}{ $issue->{filename} } }, $issue->{message};
        }
        else {
            push @{ $self->{issues} }, $issue->{message};
        }
    }

    return;
}

1;

__END__

=pod

=head1 NAME

Git::Lint - lint git commits and messages

=head1 SYNOPSIS

 use Git::Lint;

 my $lint = Git::Lint->new();
 $lint->run({ check => 'commit', profile => 'default' });
 $lint->run({ check => 'message', file => 'file_path', profile => 'default' });

 git-lint [--check commit] [--check message <message_file>]
          [--profile <name>]
          [--version] [--help]

=head1 DESCRIPTION

C<Git::Lint> is a pluggable framework for linting git commits and messages.

For the commandline interface to C<Git::Lint>, see the documentation for L<git-lint>.

For adding check modules, see the documentation for L<Git::Lint::Check::Commit> and L<Git::Lint::Check::Message>.

=head1 CONSTRUCTOR

=over

=item new

Returns a reference to a new C<Git::Lint> object.

=back

=head1 METHODS

=over

=item run

Loads the check modules as defined by C<profile>.

C<run> expects the following arguments:

B<profile>

The name of a defined set of check modules to run.

B<check>

Either C<commit> or C<message>.

B<file>

If C<check> is C<message>, C<file> is required.

=item config

Returns the L<Git::Lint::Config> object created by C<Git::Lint>.

=back

=head1 INSTALLATION

To install C<Git::Lint>, download the latest release, then extract.

 tar xzvf Git-Lint-0.008.tar.gz
 cd Git-Lint-0.008

or clone the repo.

 git clone https://github.com/renderorange/Git-Lint.git
 cd Git-Lint

Generate the build and installation tooling.

 perl Makefile.PL

Then build, test, and install.

 make
 make test && make install

C<Git::Lint> can also be installed using L<cpanm>.

 cpanm Git::Lint

=head1 CONFIGURATION

Configuration is done through C<git config> files (F<~/.gitconfig> or F</repo/.git/config>).

Only one profile, C<default>, is defined internally. C<default> contains all check modules by default.

The C<default> profile can be overridden through C<git config> files (F<~/.gitconfig> or F</repo/.git/config>).

To set the default profile to only run the C<Whitespace> commit check:

 [lint "profiles.commit"]
     default = Whitespace

Or set the default profile to C<Whitespace> and the fictional commit check, C<Flipdoozler>:

 [lint "profiles.commit"]
     default = Whitespace, Flipdoozler

Additional profiles can be added with a new name and list of checks to run.

 [lint "profiles.commit"]
     default = Whitespace, Flipdoozler
     hardcore = Other, Module, Names

Message check profiles can also be defined.

 [lint "profiles.message"]
     # override the default profile to only contain SummaryLength, SummaryEndingPeriod, and BlankLineAfterSummary
     default = SummaryLength, SummaryEndingPeriod, BlankLineAfterSummary
     # create a summary profile with specific modules
     summary = SummaryEndingPeriod, SummaryLength

An example configuration is provided in the C<examples> directory of this project.

Configuration is required.  If no configuration exists, an error will be printed to STDERR, but the action allowed to complete.

 blaine@base ~/git/test (master *) $ git add test
 blaine@base ~/git/test (master +) $ git commit
 git-lint: [error] configuration setup is required. see the documentation for instructions.
 [master 894b6d0] test
  1 file changed, 1 insertion(+), 1 deletion(-)
 blaine@base ~/git/test (master) $

=head1 ADDING NEW CHECK MODULES

C<git-lint> can be configured to load check modules from a local directory using the C<localdir> configuration setting.

To load modules from a local directory, add the lint C<config> setting, with C<localdir> key and directory location to the git config file.

 [lint "config"]
     localdir = /home/blaine/tmp/git-lint/lib

In this example, we're adding a new commit check, C<Flipdoozler>.  Create the local directory and path for the new module.

 $ mkdir -p /home/blaine/tmp/git-lint/lib/Git/Lint/Check/Commit

Then add the new check module.

 $ vi /home/blaine/tmp/git-lint/lib/Git/Lint/Check/Commit/Flipdoozler.pm
 package Git::Lint::Check::Commit::Flipdoozler;
 ...

Update the commit check profile to use the new module.

 [lint "profiles.commit"]
     default = Whitespace, IndentTabs, MixedIndentTabsSpaces, Flipdoozler

C<git-lint> will now warn for the check contained in Flipdoozler.

 blaine@base ~/git/test (master +) $ git commit
 git-lint: [commit] test - Flipdoozler (line 18)
 blaine@base ~/git/test (master +) $

=head1 ENABLING CHECKS FOR REPOS

To enable as a C<pre-commit> hook, copy the C<pre-commit> script from the C<example/hooks> directory into the C<.git/hooks> directory of the repo you want to check.

Once copied, update the path and options to match your path and preferred profile.

To enable as a C<commit-msg> hook, copy the C<commit-msg> script from the C<example/hooks> directory into the C<.git/hooks> directory of the repo you want to check.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2022 Blaine Motsinger under the MIT license.

=head1 AUTHOR

Blaine Motsinger C<blaine@renderorange.com>

=cut
